server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/MailToAllUsers.java

```diff
@@ -0,0 +1,116 @@
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.transport.mailets;
+
+import java.util.Optional;
+
+import jakarta.inject.Inject;
+import jakarta.mail.MessagingException;
+
+import org.apache.james.core.MailAddress;
+import org.apache.james.core.Username;
+import org.apache.james.lifecycle.api.LifecycleUtil;
+import org.apache.james.user.api.UsersRepository;
+import org.apache.mailet.Mail;
+import org.apache.mailet.base.GenericMailet;
+
+import com.github.fge.lambdas.Throwing;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+/**
+ * A mailet that helps to email all users in the system. The emails are sent in batches to manage
+ * the load. The first batch is sent directly, while the remaining batches are sent asynchronously.
+ * The batch size can be configured via the <b>batchSize</b> parameter (optional, defaults to 100).
+ *
+ * <h3>Configuration</h3>
+ * <pre><code>
+ * {@code
+ * <matcher name="notify-matcher" match="org.apache.james.mailetcontainer.impl.matchers.And">
+ *     <matcher match="SenderIs=admin@gov.org"/>
+ *     <matcher match="RecipientIs=all@gov.org"/>
+ * </matcher>
+ * <mailet match="notify-matcher" class="MailToAllUsers">
+ *     <batchSize>100</batchSize>
+ * </mailet>
+ * }
+ * </code></pre>
+ *
+ */
+public class MailToAllUsers extends GenericMailet {
+    private static final int DEFAULT_BATCH_SIZE = 100;
+    private final UsersRepository usersRepository;
+    private int batchSize;
+
+    @Inject
+    public MailToAllUsers(UsersRepository usersRepository) {
+        this.usersRepository = usersRepository;
+    }
+
+    @Override
+    public void init() throws MessagingException {
+        batchSize = Integer.parseInt(Optional.ofNullable(getInitParameter("batchSize"))
+            .orElse(String.valueOf(DEFAULT_BATCH_SIZE)));
+    }
+
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        Flux<Flux<MailAddress>> recipientsBatches = Flux.from(usersRepository.listReactive())
+            .map(Throwing.function(Username::asMailAddress))
+            .window(batchSize);
+
+        Flux.merge(sendMailToFirstRecipientsBatchDirectly(mail, recipientsBatches),
+                sendMailToRemainingRecipientsBatchAsynchronously(mail, recipientsBatches))
```

Comment:
The issue is that we subscribe twice and the order might not be stable.

This means the first 100 items on the first subscriptions are not necessarily the ones of the next subscriptions.

How about:

```
Flux.from(usersRepository.listReactive())
            .map(Throwing.function(Username::asMailAddress))
            .window(batchSize);
            .index()
            .flatMap(tuple -> if (tuple.key() == 0) {
                return sendMailToFirstRecipientsBatchDirectly(mail, tuple.value());
            } else {
                return sendMailToRemainingRecipientsBatchAsynchronously(mail, tuple.value());
            }).then().block();

Reply:
We may use `Flux.concat` as it would wait for the first Publisher to finish.

Or `sendMailToFirstRecipientsBatchDirectly.then(sendMailToRemainingRecipientsBatchAsynchronously)`.

Reply:
BTW `recipientsBatches`  is a cold publisher so is it supposed to produce data isolated regardless of subscribers?

Reply:
Yes but because it is subscribed twice data may defer.

Reply:
first call: `a, b, c, d, e` => send mail synchronously to `a, b, c`

second call `d, e, b, a, c` => send mail asynchrounously to `a, c`

 - We called twice the ldap
 - and we sent `a, b, c, a, c` and not `a, b, c, d, e`